src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java

```diff
@@ -362,23 +369,50 @@ private PaymentTransactionInfoPlugin executeInitialTransaction(final Transaction
+            killbillMetrics.recordHistogramValue(transactionType + ".amount", amountInCents);
+        } else {
+            amountInCents = null;
+        }
 
-        final PurchaseResult response = transactionExecutor.execute(transactionAmount, paymentData, orderData, userData, termUrl, splitSettlementData);
+        final AdyenPaymentTransactionInfoPlugin paymentTransaction;
         try {
-            dao.addResponse(kbAccountId, kbPaymentId, kbTransactionId, transactionType, amount, currency, response, utcNow, context.getTenantId());
-            return new AdyenPaymentTransactionInfoPlugin(kbPaymentId, kbTransactionId, transactionType, amount, currency, utcNow, response);
-        } catch (final SQLException e) {
-            throw new PaymentPluginApiException("Payment went through, but we encountered a database error. Payment details: " + response.toString(), e);
+            final Account account = getAccount(kbAccountId, context);
+
+            final Long transactionAmount = (amount == null ? null : amount.longValue());
+            final PaymentData paymentData = buildPaymentData(account, kbPaymentId, kbTransactionId, kbPaymentMethodId, currency, properties, context);
+            final OrderData orderData = buildOrderData(account, properties);
+            final UserData userData = buildUserData(account, properties);
+            final String termUrl = PluginProperties.findPluginPropertyValue(PROPERTY_TERM_URL, properties);
+            final SplitSettlementData splitSettlementData = null;
+            final DateTime utcNow = clock.getUTCNow();
+
+            final PurchaseResult response = transactionExecutor.execute(transactionAmount, paymentData, orderData, userData, termUrl, splitSettlementData);
+            try {
+                dao.addResponse(kbAccountId, kbPaymentId, kbTransactionId, transactionType, amount, currency, response, utcNow, context.getTenantId());
+                paymentTransaction = new AdyenPaymentTransactionInfoPlugin(kbPaymentId, kbTransactionId, transactionType, amount, currency, utcNow, response);
+            } catch (final SQLException e) {
+                throw new PaymentPluginApiException("Payment went through, but we encountered a database error. Payment details: " + response.toString(), e);
+            }
+            killbillMetrics.markMeter(transactionType + "." + paymentTransaction.getStatus());
+            if (amountInCents != null) {
+                killbillMetrics.recordHistogramValue(transactionType + "." + paymentTransaction.getStatus() + ".amount", amountInCents);
+            }
+        } catch(Exception e) {
+            killbillMetrics.markMeter(transactionType + ".exception");
+            if (amountInCents != null) {
+                killbillMetrics.recordHistogramValue(transactionType + ".exception.amount", amountInCents);
+            }
+            if (e instanceof PaymentPluginApiException) {
+                throw (PaymentPluginApiException) e;
```

Comment:
I think it's a good idea to send metrics about exception, but for that we need to improve the error handling and be able to get a better idea of what actually failed. Actually, the same applies to declined transactions and ERROR transaction statuses.

I think this error handling is something we can improve, but I didn't dive into that since, for now it's out of the scope for this.